### PR TITLE
storage: use MVCCBlindPut for the raft-last-index key

### PR DESCRIPTION
**Pull request by petermattis**
Use a blind put when persisting the raft-last-index key in order to
avoid reading the previous value unnecessarily.

See #8640.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9263)

### CODE REVIEW

**Comment by petermattis** [comment]
Noticed while looking at a cpu profile on gamma. Not stability per-se, so perhaps this should go to develop. But fairly innocuous given that we usually use blind puts for raft log entries.

**Comment by petermattis** [comment]
Cc @cockroachdb/stability

File: storage/replica_state.go [hunk 1]
```diff
@@ -328,7 +328,10 @@ func setLastIndex(
 	var value roachpb.Value
 	value.SetInt(int64(lastIndex))
 
-	return engine.MVCCPut(ctx, eng, nil, keys.RaftLastIndexKey(rangeID),
+	// NB: It is safe to use a blind put here because the raft-last-index key is
+	// is not part of the replica MVCC stats computation. See also that we're
```
Comment (spencerkimball) [inline storage/replica_state.go]: s/is is/is/

**Comment by spencerkimball** [comment]
LGTM. I'm fine merging this to master. How much of a blip is this on performance?

File: storage/replica_state.go (hunk 1 above)
Comment (petermattis) [inline storage/replica_state.go]: Done.

**Comment by petermattis** [comment]
Hard to say how much effect this will have. I'd expect a few percent. I'm running a local `block_writer` test write now.

File: storage/replica_state.go [hunk 2]
```diff
@@ -328,7 +328,10 @@ func setLastIndex(
 	var value roachpb.Value
 	value.SetInt(int64(lastIndex))
 
-	return engine.MVCCPut(ctx, eng, nil, keys.RaftLastIndexKey(rangeID),
+	// NB: It is safe to use a blind put here because the raft-last-index key is
+	// not part of the replica MVCC stats computation. See also that we're
+	// passing nil for the MVCCStats parameter.
+	return engine.MVCCBlindPut(ctx, eng, nil /* ms */, keys.RaftLastIndexKey(rangeID),
```
Comment (tschottdorf) [inline storage/replica_state.go]: Can't we make this optimization more generally? If you're doing an inline (zero-timestamp) put (without a valueFunc) then `mvccPutInternal` (or wherever the magic happens) should be able to see whether the outcome can change the stats and can blind put.
Not saying we should do that, just curious whether I'm overlooking something.

**Comment by tschottdorf** [comment]
LGTM but I hope this isn't the starting signal for all sorts of random "innocuous" micro-optimizations. Still too early for those.

File: storage/replica_state.go (hunk 2 above)
Comment (petermattis) [inline storage/replica_state.go]: Yeah, I considered that, but I believe this and `setHardState` are the only place the optimization applies and it seemed safer to apply it only here.

**Comment by bdarnell** [comment]
LGTM, but I'd say it should go to `develop`. 

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica_state.go, line 334 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9263#-KRKZFpDZ2e3z1Qk41SR-r2-334:-KRRzBxk1G7chQz21ES5:b-ysvv83) ([raw file](https://github.com/cockroachdb/cockroach/blob/61bb674379520d90849474e44435017bd68412de/storage/replica_state.go#L334)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yeah, I considered that, but I believe this and `setHardState` are the only place the optimization applies and it seemed safer to apply it only here.
</details>
There are more, although they're not as important for performance (setReplicaDestroyedError, several timestamps). Any of the keys that use MakeRangeIDUnreplicatedKey should be able to be a blind put (except for the log itself, for which we explicitly track stats even though it is unreplicated).

The two applied index keys are counted in the stats, although if we could stop doing that they'd be able to be blind puts as well (and they're updated frequently).  

For things like this I'd rather have the MVCC code know when this optimization is safe and apply it automatically when possible than to decide on a case-by-case basis. (the use of blind puts for KV data will have to be decided at a higher level, but when the rules are as simple as this we can build it in to the lower level). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9263#-:-KRS-Man2hQKGkxBpLhj:biwh5g5)*

**Comment by petermattis** [comment]
Heh, simply change the target branch to `develop` created a mess. Let me fix.

**Comment by petermattis** [comment]
Ok, re-targeted this PR at `develop`. Should be good to go.

**Comment by petermattis** [comment]
PS Other reasonably useful places this optimization applies: `setHardState` and `MVCCSetRangeStats`.

**Comment by bdarnell** [comment]
LGTM. Reviewable looks very confused by the fact that this PR now targets develop, but the commit on github looks fine.

**Pull request merged**